libavformat/rtcenc.c

```diff
@@ -0,0 +1,2211 @@
+    /* The resource URL returned in the Location header of WHIP HTTP response. */
+    char *whip_resource_url;
+
+    /* The DTLS context. */
+    DTLSContext dtls_ctx;
+
+    /* The SRTP send context, to encrypt outgoing packets. */
+    struct SRTPContext srtp_audio_send;
+    struct SRTPContext srtp_video_send;
+    struct SRTPContext srtp_rtcp_send;
+    /* The SRTP receive context, to decrypt incoming packets. */
+    struct SRTPContext srtp_recv;
+
+    /* The time jitter base for audio OPUS stream. */
+    int64_t audio_jitter_base;
+
+    /* The UDP transport is used for delivering ICE, DTLS and SRTP packets. */
+    URLContext *udp_uc;
+
+    /* The maximum number of retries for ICE transmission. */
+    int ice_arq_max;
+    /* The step start timeout in ms for ICE transmission. */
+    int ice_arq_timeout;
+    /* The maximum number of retries for DTLS transmission. */
+    int dtls_arq_max;
+    /* The step start timeout in ms for DTLS transmission. */
+    int dtls_arq_timeout;
+    /* The size of RTP packet, should generally be set to MTU. */
+    int pkt_size;
+} RTCContext;
+
+static int on_rtp_write_packet(void *opaque, uint8_t *buf, int buf_size);
+
+/**
+ * Initialize and check the options for the WebRTC muxer.
+ */
+static av_cold int whip_init(AVFormatContext *s)
+{
+    int ret, ideal_pkt_size = 532;
+    RTCContext *rtc = s->priv_data;
+
+    /* Use the same logging context as AV format. */
+    rtc->dtls_ctx.log_avcl = s;
+    rtc->dtls_ctx.dtls_arq_max = rtc->dtls_arq_max;
+    rtc->dtls_ctx.dtls_arq_timeout = rtc->dtls_arq_timeout;
+    rtc->dtls_ctx.pkt_size = rtc->pkt_size;
+
+    if ((ret = dtls_context_init(&rtc->dtls_ctx)) < 0) {
+        av_log(s, AV_LOG_ERROR, "WHIP: Failed to init DTLS context\n");
+        return ret;
+    }
+
+    av_log(s, AV_LOG_INFO, "WHIP: Init ice_arq_max=%d, ice_arq_timeout=%d, dtls_arq_max=%d, dtls_arq_timeout=%d pkt_size=%d\n",
+        rtc->ice_arq_max, rtc->ice_arq_timeout, rtc->dtls_arq_max, rtc->dtls_arq_timeout, rtc->pkt_size);
+
+    if (rtc->pkt_size < ideal_pkt_size)
+        av_log(s, AV_LOG_WARNING, "WHIP: pkt_size=%d(<%d) is too small, may cause packet loss\n",
+            rtc->pkt_size, ideal_pkt_size);
+
+    return 0;
+}
+
+/**
+ * Parses the ISOM AVCC format of extradata and extracts SPS/PPS.
+ *
+ * This function is used to parse SPS/PPS from the extradata in ISOM AVCC format.
+ * It can handle both ISOM and annexb formats but only parses data in ISOM format.
+ * If the extradata is in annexb format, this function ignores it, and uses the entire
+ * extradata as a sequence header with SPS/PPS. Refer to ff_isom_write_avcc.
+ *
+ * @param s                Pointer to the AVFormatContext
+ * @param extradata        Pointer to the extradata
+ * @param extradata_size   Size of the extradata
+ * @returns                Returns 0 if successful or AVERROR_xxx in case of an error.
+ */
+static int isom_read_avcc(AVFormatContext *s, uint8_t *extradata, int  extradata_size)
```

Comment:
Can libavformat/avc.c replace this function?

Reply:
In avc.c, there is only a function ff_isom_write_avcc to encode avcc in ISOM format, but we need to decode it.

I didn't find a similar API to decode SPS/PPS from extradata, so we must parse it ourselves.